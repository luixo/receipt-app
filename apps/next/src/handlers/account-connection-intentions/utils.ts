import * as trpc from "@trpc/server";

import { Database } from "next-app/db";
import { ACCOUNT_CONNECTIONS_INTENTIONS } from "next-app/db/consts";
import { AccountsId, UsersId } from "next-app/db/models";
import { getUserById } from "next-app/handlers/users/utils";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: string,
	asUserId: UsersId
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select("id")
		.where("email", "=", toEmail)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Account with email ${toEmail} does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId)
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new trpc.TRPCError({
			code: "CONFLICT",
			message: `Account with email ${toEmail} is already connected to user ${connectedUser.name}.`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where("accountId", "=", targetAccount.id)
		.where("targetAccountId", "=", fromAccountId)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where("ownerAccountId", "=", targetAccount.id)
				.where("id", "=", viceVersaIntention.userId)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where("ownerAccountId", "=", fromAccountId)
				.where("id", "=", asUserId)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where("accountId", "=", targetAccount.id)
				.where("targetAccountId", "=", fromAccountId)
				.executeTakeFirst();
		});
		return {
			id: targetAccount.id,
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				created: new Date(),
			})
			.executeTakeFirst();
		return {
			id: targetAccount.id,
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where("accountId", "=", fromAccountId)
				.where("targetAccountId", "=", targetAccount.id)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await getUserById(
				database,
				existingIntention.userId,
				["name"]
			);
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to ${toEmail} as user ${
					existingUser!.name
				}.`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user ${user.name}.`,
			});
		}
		throw e;
	}
};

export const removeIntention = async (
	database: Database,
	intention:
		| {
				targetAccountId: AccountsId;
				accountId: AccountsId;
		  }
		| undefined,
	intentionType: string
) => {
	if (!intention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Intention ${intentionType} does not exist.`,
		});
	}
	await database
		.deleteFrom("accountConnectionsIntentions")
		.where("accountId", "=", intention.accountId)
		.where("targetAccountId", "=", intention.targetAccountId)
		.execute();
};
