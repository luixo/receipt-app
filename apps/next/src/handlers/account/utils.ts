import { Selection } from "kysely";
import { Database, ReceiptsSelectExpression, ReceiptsDatabase } from "../../db";
import { AccountsId } from "../../db/models";

export const getAccountById = <SE extends ReceiptsSelectExpression<"accounts">>(
	database: Database,
	id: AccountsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("id", "=", id)
		.executeTakeFirst();
};

export const getAccountByEmail = <
	SE extends ReceiptsSelectExpression<"accounts">
>(
	database: Database,
	email: string,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("email", "=", email)
		.executeTakeFirst();
};
