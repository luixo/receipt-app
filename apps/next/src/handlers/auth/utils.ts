import * as trpc from "@trpc/server";

import { DAY } from "app/utils/time";
import type { Database } from "next-app/db";
import type { AccountsId, SessionsSessionId } from "next-app/db/models";
import { generateConfirmEmailEmail } from "next-app/email/utils";
import { getUuid } from "next-app/utils/crypto";
import { getEmailClient, isEmailServiceActive } from "next-app/utils/email";

// How long a session should last
const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long before the session expires should we update the expiration date
const SESSION_SHOULD_UPDATE_DURATION = 28 * DAY;

const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);
export const shouldUpdateExpirationDate = (expirationTimestamp: Date) =>
	expirationTimestamp.valueOf() - Date.now() < SESSION_SHOULD_UPDATE_DURATION;

export const createAuthorizationSession = async (
	database: Database,
	accountId: AccountsId,
) => {
	const uuid = getUuid();
	const expirationDate = getExpirationDate();
	await database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const updateAuthorizationSession = async (
	database: Database,
	authToken: SessionsSessionId,
) => {
	database
		.updateTable("sessions")
		.set({ expirationTimestamp: getExpirationDate() })
		.where("sessionId", "=", authToken)
		.executeTakeFirst();
};

export const removeAuthorizationSession = async (
	database: Database,
	sessionId: SessionsSessionId,
) => {
	await database
		.deleteFrom("sessions")
		.where("sessionId", "=", sessionId)
		.executeTakeFirst();
};

export const sendVerificationEmail = async (email: string, token: string) => {
	if (!isEmailServiceActive()) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient().send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(token),
		});
	} catch (e) {
		throw new trpc.TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
