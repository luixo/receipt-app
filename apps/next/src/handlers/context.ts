import * as trpcNext from "@trpc/server/adapters/next";
import { NextApiRequest, NextApiResponse } from "next";
import { Logger, logger } from "../utils/logger";

export type Context = {
	req: NextApiRequest;
	res: NextApiResponse;
	logger: Logger;
	debug: boolean;
};

export const createContext = (
	opts: trpcNext.CreateNextContextOptions
): Context => ({
	req: opts.req,
	res: opts.res,
	logger,
	debug: Boolean(opts.req.headers.debug),
});
