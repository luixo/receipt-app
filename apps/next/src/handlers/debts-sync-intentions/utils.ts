import * as trpc from "@trpc/server";
import { z } from "zod";

import { Database } from "next-app/db";
import { AccountsId, DebtsId } from "next-app/db/models";
import { getValidParticipants } from "next-app/handlers/receipt-items/utils";

export const getDebtIntention = async (
	database: Database,
	debtId: DebtsId,
	accountId: AccountsId,
) =>
	database
		.selectFrom("debts")
		.where("debts.id", "=", debtId)
		.leftJoin("debts as debtsMine", (qb) =>
			qb
				.onRef("debts.id", "=", "debtsMine.id")
				.on("debtsMine.ownerAccountId", "=", accountId),
		)
		.leftJoin("debtsSyncIntentions", (qb) =>
			qb.onRef("debts.id", "=", "debtsSyncIntentions.debtId"),
		)
		.leftJoin("debts as debtsTheir", (qb) =>
			qb
				.onRef("debts.id", "=", "debtsTheir.id")
				.on("debtsTheir.ownerAccountId", "<>", accountId),
		)
		.select([
			"debtsSyncIntentions.ownerAccountId as intentionAccountId",
			"debtsMine.lockedTimestamp as mineLockedTimestamp",
			"debtsMine.ownerAccountId as mineOwnerAccountId",
			"debtsTheir.lockedTimestamp as theirLockedTimestamp",
			"debtsTheir.ownerAccountId as theirOwnerAccountId",
		])
		.executeTakeFirst();

const syncStatusSchema = z.discriminatedUnion("type", [
	// The foreign user doesn't have a debt
	z.strictObject({ type: z.literal("nosync") }),
	// The foreign user debt's data is different
	z.strictObject({
		type: z.literal("unsync"),
		intention: z
			.strictObject({
				direction: z.union([z.literal("self"), z.literal("remote")]),
			})
			.optional(),
	}),
	// The foreign user debt's data is in sync
	z.strictObject({ type: z.literal("sync") }),
]);

export type SyncStatus = z.infer<typeof syncStatusSchema>;

export const getSyncStatus = async (
	database: Database,
	debtId: DebtsId,
	accountId: AccountsId,
): Promise<SyncStatus> => {
	const debtIntention = await getDebtIntention(database, debtId, accountId);
	if (!debtIntention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Debt ${debtId} does not exist.`,
		});
	}
	const {
		mineOwnerAccountId,
		mineLockedTimestamp,
		theirOwnerAccountId,
		theirLockedTimestamp,
		intentionAccountId,
	} = debtIntention;
	if (!mineOwnerAccountId) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: `Debt ${debtId} does not exist for user ${accountId}.`,
		});
	}
	if (
		theirLockedTimestamp &&
		mineLockedTimestamp &&
		theirLockedTimestamp.valueOf() === mineLockedTimestamp.valueOf()
	) {
		return { type: "sync" };
	}
	if (!intentionAccountId) {
		if (theirOwnerAccountId) {
			return { type: "unsync" };
		}
		return { type: "nosync" };
	}
	if (intentionAccountId === accountId) {
		return {
			type: "unsync",
			intention: { direction: "self" },
		};
	}
	return {
		type: "unsync",
		intention: { direction: "remote" },
	};
};

type Participant = Awaited<ReturnType<typeof getValidParticipants>>[number];

export const upsertDebtIntentionFromReceipt = async (
	database: Database,
	participants: Participant[],
	ownerAccountId: AccountsId,
	created: Date,
) =>
	database
		.insertInto("debtsSyncIntentions")
		.values(
			participants.map((participant) => ({
				debtId: participant.debtId,
				ownerAccountId,
				lockedTimestamp: created,
			})),
		)
		.onConflict((oc) =>
			oc.column("debtId").doUpdateSet({
				ownerAccountId: (eb) => eb.ref("excluded.ownerAccountId"),
				lockedTimestamp: (eb) => eb.ref("excluded.lockedTimestamp"),
			}),
		)
		.execute();
