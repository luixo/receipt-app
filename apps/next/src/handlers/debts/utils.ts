import { Selection } from "kysely";

import { ReceiptsSelectExpression, Database } from "next-app/db";
import { DEBTS } from "next-app/db/consts";
import { AccountsId, DebtsId, Receipts } from "next-app/db/models";
import { ReceiptsDatabase } from "next-app/db/types";
import { getValidParticipants } from "next-app/handlers/receipt-items/utils";

export const getDebt = <SE extends ReceiptsSelectExpression<"debts">>(
	database: Database,
	id: DebtsId,
	ownerAccountId: AccountsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "debts", SE> | undefined> =>
	database
		.selectFrom("debts")
		.where("id", "=", id)
		.where("ownerAccountId", "=", ownerAccountId)
		.select(selectExpression)
		.executeTakeFirst();

const getReceiptDebtName = (receipt: Pick<Receipts, "name">) =>
	`Receipt "${receipt.name}"`;

type Participant = Awaited<ReturnType<typeof getValidParticipants>>[number];

export const upsertDebtFromReceipt = async (
	database: Database,
	participants: Participant[],
	receipt: Pick<
		Receipts,
		"id" | "ownerAccountId" | "lockedTimestamp" | "name" | "issued" | "currency"
	>,
	created: Date
) =>
	database
		.insertInto("debts")
		.values(
			participants.map((participant) => ({
				id: participant.debtId,
				ownerAccountId: receipt.ownerAccountId,
				userId: participant.remoteUserId,
				note: getReceiptDebtName(receipt),
				currency: receipt.currency,
				created,
				timestamp: receipt.issued,
				amount: participant.sum.toString(),
				receiptId: receipt.id,
				lockedTimestamp: receipt.lockedTimestamp,
			}))
		)
		.onConflict((oc) =>
			oc
				.constraint(DEBTS.CONSTRAINTS.OWNER_ID_RECEIPT_ID_USER_ID_TUPLE)
				.doUpdateSet({
					currency: (eb) => eb.ref("excluded.currency"),
					timestamp: (eb) => eb.ref("excluded.timestamp"),
					amount: (eb) => eb.ref("excluded.amount"),
					lockedTimestamp: (eb) => eb.ref("excluded.lockedTimestamp"),
				})
		)
		.returning(["id as debtId", "userId", "note"])
		.execute();

export const getDebtsResult = (
	participants: Participant[],
	actualDebts: Awaited<ReturnType<typeof upsertDebtFromReceipt>>,
	receipt: Pick<
		Receipts,
		"id" | "ownerAccountId" | "lockedTimestamp" | "name" | "issued" | "currency"
	>,
	created: Date
) =>
	participants.map((participant) => {
		const actualDebt = actualDebts.find(
			(debt) => debt.userId === participant.remoteUserId
		)!;
		return {
			debtId: actualDebt.debtId,
			userId: participant.remoteUserId,
			updated: actualDebt.debtId !== participant.debtId,
			note: actualDebt.note,
			currency: receipt.currency,
			created,
			timestamp: receipt.issued,
			amount: Number(participant.sum),
		};
	});
