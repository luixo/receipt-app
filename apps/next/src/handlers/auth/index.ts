import * as trpc from "@trpc/server";

import { UnauthorizedContext } from "next-app/handlers/context";

import { router as loginRouter } from "./login";
import { router as registerRouter } from "./register";
import { router as resetPasswordRouter } from "./reset-password";

export const router = trpc
	.router<UnauthorizedContext>()
	.merge(loginRouter)
	.merge(registerRouter)
	.merge(resetPasswordRouter);
