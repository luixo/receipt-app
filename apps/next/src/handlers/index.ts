import * as trpc from "@trpc/server";
import { Context } from "./context";
import { middleware as loggerMiddleware } from "./middlewares/logger";

import { router as receiptsRouter } from "./receipts/index";

export const router = trpc
	.router<Context>()
	.middleware(loggerMiddleware)
	.merge("receipts.", receiptsRouter);
