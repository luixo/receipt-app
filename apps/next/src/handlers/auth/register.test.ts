import { faker } from "@faker-js/faker";
import { describe, expect } from "vitest";

import {
	MAX_PASSWORD_LENGTH,
	MAX_USERNAME_LENGTH,
	MIN_PASSWORD_LENGTH,
	MIN_USERNAME_LENGTH,
} from "app/utils/validation";
import { router } from "next-app/handlers/index";
import { UUID_REGEX } from "next-app/handlers/validation";
import { createContext } from "next-tests/utils/context";
import { insertAccountWithSession } from "next-tests/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
} from "next-tests/utils/expect";
import { test } from "next-tests/utils/test";

describe("auth.register", () => {
	describe("input verification", () => {
		describe("email", () => {
			test("invalid", async ({ ctx }) => {
				const caller = router.createCaller(createContext(ctx));
				await expectTRPCError(
					() =>
						caller.auth.register({
							email: "invalid@@mail.org",
							password: "a".repeat(MIN_PASSWORD_LENGTH),
							name: "a".repeat(MIN_USERNAME_LENGTH),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "email": Invalid email address`,
				);
			});
		});

		describe("password", () => {
			test("minimal length", async ({ ctx }) => {
				const caller = router.createCaller(createContext(ctx));
				await expectTRPCError(
					() =>
						caller.auth.register({
							email: "valid@mail.org",
							password: "a".repeat(MIN_PASSWORD_LENGTH - 1),
							name: "a".repeat(MIN_USERNAME_LENGTH),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "password": Minimal length for password is ${MIN_PASSWORD_LENGTH}`,
				);
			});

			test("maximum length", async ({ ctx }) => {
				const caller = router.createCaller(createContext(ctx));
				await expectTRPCError(
					() =>
						caller.auth.register({
							email: "valid@mail.org",
							password: "a".repeat(MAX_PASSWORD_LENGTH + 1),
							name: "a".repeat(MIN_USERNAME_LENGTH),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "password": Maximum length for password is ${MAX_PASSWORD_LENGTH}`,
				);
			});
		});

		describe("name", () => {
			test("minimal length", async ({ ctx }) => {
				const caller = router.createCaller(createContext(ctx));
				await expectTRPCError(
					() =>
						caller.auth.register({
							email: "valid@mail.org",
							password: "a".repeat(MIN_PASSWORD_LENGTH),
							name: "a".repeat(MIN_USERNAME_LENGTH - 1),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "name": Minimal length for user name is ${MIN_USERNAME_LENGTH}`,
				);
			});

			test("maximum length", async ({ ctx }) => {
				const caller = router.createCaller(createContext(ctx));
				await expectTRPCError(
					() =>
						caller.auth.register({
							email: "valid@mail.org",
							password: "a".repeat(MIN_PASSWORD_LENGTH),
							name: "a".repeat(MAX_USERNAME_LENGTH + 1),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "name": Maximum length for user name is ${MAX_USERNAME_LENGTH}`,
				);
			});
		});

		test("email already exist", async ({ ctx }) => {
			const caller = router.createCaller(createContext(ctx));
			const {
				account: { email: existingEmail },
			} = await insertAccountWithSession(ctx);
			await expectTRPCError(
				() =>
					caller.auth.register({
						email: existingEmail,
						password: "a".repeat(MIN_PASSWORD_LENGTH),
						name: "a".repeat(MIN_USERNAME_LENGTH),
					}),
				"CONFLICT",
				"Email already exist",
			);
		});
	});

	describe("functionality", () => {
		test("register successful", async ({ ctx }) => {
			ctx.emailOptions.active = false;
			const context = createContext(ctx);
			const caller = router.createCaller(context);
			await expectDatabaseDiffSnapshot(ctx, async () => {
				const result = await caller.auth.register({
					email: faker.internet.email(),
					password: faker.internet.password(),
					name: faker.person.firstName(),
				});
				expect(result.account.id).toMatch(UUID_REGEX);
				expect(result).toEqual<typeof result>({
					account: { id: result.account.id },
				});
			});
			expect(context.setHeaders).toMatchSnapshot();
		});

		test("email sent if active", async ({ ctx }) => {
			const caller = router.createCaller(createContext(ctx));
			await expectDatabaseDiffSnapshot(ctx, () =>
				caller.auth.register({
					email: faker.internet.email(),
					password: faker.internet.password(),
					name: faker.person.firstName(),
				}),
			);
			expect(ctx.emailOptions.cachedMessages).toHaveLength(1);
			expect(ctx.emailOptions.cachedMessages![0]).toMatchSnapshot();
		});

		test("email reports error if broken", async ({ ctx }) => {
			ctx.emailOptions.broken = true;
			const context = createContext(ctx);
			const caller = router.createCaller(context);
			await expectTRPCError(
				() =>
					caller.auth.register({
						email: faker.internet.email(),
						password: faker.internet.password(),
						name: faker.person.firstName(),
					}),
				"INTERNAL_SERVER_ERROR",
				"Something went wrong: Test context broke email service error",
			);
			expect(ctx.emailOptions.cachedMessages).toHaveLength(0);
		});
	});
});
