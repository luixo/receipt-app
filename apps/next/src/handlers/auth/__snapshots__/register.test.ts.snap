// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`auth.register > functionality > email sent if active 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
+ INSERT INTO public.accounts VALUES ('9f899965-3642-4806-8e47-9628617febef', 'roberto9@yahoo.com', '747028d153a4bb69c984302eb6ff4628f1253f54aef3dd47b6770768f0d077ef00829147d9c161a307aa8c1d57591bcd31fed28be80a8a13654553b5d15cf980', 'b4cf0fd7fb1f7f92fa1c7fbceadbfb16ad9c7bc8afaceb6dc5acccdcc814f9bb6f6ea356d3ad090fda433dd7c7857b946fff91d08fd62f79edde67e6a0cc0bb2', '1668eb65-2967-4c4e-94cd-98ddf052c611', '2020-01-01 00:00:00');
+ INSERT INTO public.users VALUES ('9f899965-3642-4806-8e47-9628617febef', 'Hillary', NULL, '9f899965-3642-4806-8e47-9628617febef', true, true, '9f899965-3642-4806-8e47-9628617febef');
+ INSERT INTO public.sessions VALUES ('b7ef42fd-6010-486f-a500-5003595e03c8', '9f899965-3642-4806-8e47-9628617febef', '2020-01-31 00:00:00');"
`;

exports[`auth.register > functionality > email sent if active 2`] = `
"
	<!doctype html lang=\\"en\\">
	<html lang=\\"en\\">
		<head>
			<meta name=\\"viewport\\" content=\\"width=device-width\\" />
			<meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\" />
			<title>Receipt App</title>
			<style> body {background-color:#eaebed;font-family:sans-serif;font-size:14px;line-height:1.4;margin:0;padding:0;-webkit-font-smoothing:antialiased;-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;} h1,h2,h3,h4 {color:#06090f;font-family:sans-serif;font-weight:400;line-height:1.4;margin:0;} p {font-family:sans-serif;font-size:14px;font-weight:normal;margin:0;margin-bottom:15px;} @media only screen and (max-width: 620px) { p {font-size:16px !important;} a {font-size:16px !important;} td {font-size:16px !important;} span {font-size:16px !important;} .container {padding:0 !important;width:100% !important;} .content {padding:0 !important;} .main {border-left-width:0 !important;border-radius:0 !important;border-right-width:0 !important;} .wrapper {padding:10px !important;} table[class=body] h3 {font-size:22px !important;margin-bottom:10px !important;} .btn table {width:100% !important;} .btn a {width:100% !important;}} a {color:#ec0867;text-decoration:underline;} .container {display:block;margin:0 auto !important;max-width:580px;padding:10px;width:580px;} .content {box-sizing:border-box;display:block;margin:0 auto;max-width:580px;padding:10px;} .preheader {color:transparent;display:none;height:0;max-height:0;max-width:0;opacity:0;overflow:hidden;mso-hide:all;visibility:hidden;width:0;} table {border-collapse:separate;mso-table-lspace:0pt;mso-table-rspace:0pt;min-width:100%;width:100%;} table td {font-family:sans-serif;font-size:14px;vertical-align:top;} .header {padding:20px 0;} h1 {font-size:35px;font-weight:300;text-align:center;text-transform:capitalize;} .align-center {text-align:center;} .main {background:#ffffff;border-radius:3px;width:100%;} .wrapper {box-sizing:border-box;padding:20px;} .btn {box-sizing:border-box;width:100%;} .btn > tbody > tr > td  {padding-bottom:15px;} .btn table {min-width:auto;width:auto;} .btn table td {background-color:#ffffff;border-radius:5px;text-align:center;} .btn a {background-color:#ffffff;border:solid 1px #ec0867;border-radius:5px;box-sizing:border-box;color:#ec0867;cursor:pointer;display:inline-block;font-size:14px;font-weight:bold;margin:0;padding:12px 25px;text-decoration:none;text-transform:capitalize;} .btn-primary table td {background-color:#ec0867;} .btn-primary a {background-color:#ec0867;border-color:#ec0867;color:#ffffff;} .btn-primary table td:hover {background-color:#d5075d !important;} .btn-primary a:hover {background-color:#d5075d !important;border-color:#d5075d !important;} .footer {clear:both;margin-top:10px;text-align:center;width:100%;} .footer td,.footer p,.footer span,.footer a {color:#9a9ea6;font-size:12px;text-align:center;} .footer-block {padding-bottom:10px;padding-top:10px;}</style>
		</head>
		<body><table role=\\"presentation\\" style=\\"border:0\\" cellPadding=\\"0\\" cellSpacing=\\"0\\" class=\\"body\\"><tr><td> </td><td class=\\"container\\"><div class=\\"header\\"><table role=\\"presentation\\" style=\\"border:0\\" cellPadding=\\"0\\" cellSpacing=\\"0\\" width=\\"100%\\"><tr><td class=\\"align-center\\" width=\\"100%\\"><h1>Receipt App</h1></td></tr></table></div><div class=\\"content\\"><span class=\\"preheader\\">Receipt App confirm email</span><table role=\\"presentation\\" class=\\"main\\"><tr><td class=\\"wrapper\\"><table role=\\"presentation\\" style=\\"border:0\\" cellPadding=\\"0\\" cellSpacing=\\"0\\"><tr><td><h3>✨Welcome to Receipt App✨</h3><p>Happy counting!</p><table role=\\"presentation\\" style=\\"border:0\\" cellPadding=\\"0\\" cellSpacing=\\"0\\" class=\\"btn btn-primary\\"><tbody><tr><td align=\\"left\\"><table role=\\"presentation\\" style=\\"border:0\\" cellPadding=\\"0\\" cellSpacing=\\"0\\"><tbody><tr><td> <a href=\\"http://receipt-app.test/confirm-email?token=1668eb65-2967-4c4e-94cd-98ddf052c611\\" rel=\\"noreferrer\\" target=\\"_blank\\">Confirm email</a> </td></tr></tbody></table></td></tr></tbody></table><p>Didn’t register in Receipt App? Click below to void your account.</p><table role=\\"presentation\\" style=\\"border:0\\" cellPadding=\\"0\\" cellSpacing=\\"0\\" class=\\"btn btn-primary\\"><tbody><tr><td align=\\"left\\"><table role=\\"presentation\\" style=\\"border:0\\" cellPadding=\\"0\\" cellSpacing=\\"0\\"><tbody><tr><td> <a href=\\"http://receipt-app.test/void-account?token=1668eb65-2967-4c4e-94cd-98ddf052c611\\" rel=\\"noreferrer\\" target=\\"_blank\\">Void account</a> </td></tr></tbody></table></td></tr></tbody></table></td></tr></table></td></tr></table><div class=\\"footer\\"><table role=\\"presentation\\" style=\\"border:0\\" cellPadding=\\"0\\" cellSpacing=\\"0\\"><tr><td class=\\"footer-block\\"><span class=\\"apple-link\\">Contact us on receipt@luixo.ru</span><br/>You can click <a href=\\"http://receipt-app.test/unsubscribe\\">unsubscribe</a>, but there is nothing to unsubscribe from.</td></tr><tr><td class=\\"footer-block powered-by\\">Inspired by <a href=\\"https://postdrop.io\\">Postdrop</a>.</td></tr></table></div></div></td><td> </td></tr></table>
		</body>
	</html>"
`;

exports[`auth.register > functionality > register successful 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
+ INSERT INTO public.accounts VALUES ('f6c4adc4-b35f-40f4-91ac-a439e3b74466', 'octavia.schuster@hotmail.com', '05ee93383ba895e6a7c28c5f1851969e7a30a7459c3763b6ac167997b7e164b84558dbdc5058a72d7cb2a2c0acdbb3338b906e3b0b70fd77ced272a6d8541f8a', 'f3ce5da5bca27c69d47cddacaf30e0da6c11b89779ffcf50d207fd1c9cccc09360def24f1b08692eadba855ad7a292addd56b9a65f26b1ab8dfd1f20fb32ba4b', NULL, NULL);
+ INSERT INTO public.users VALUES ('f6c4adc4-b35f-40f4-91ac-a439e3b74466', 'Destinee', NULL, 'f6c4adc4-b35f-40f4-91ac-a439e3b74466', true, true, 'f6c4adc4-b35f-40f4-91ac-a439e3b74466');
+ INSERT INTO public.sessions VALUES ('dffcbe65-285f-4554-8669-20df32ae8838', 'f6c4adc4-b35f-40f4-91ac-a439e3b74466', '2020-01-31 00:00:00');"
`;
